File: backend/app/schemas/employee.py
# E:\Audit\AMS\backend\app\schemas\employee.py

import re
from datetime import date, datetime
from pydantic import BaseModel, EmailStr, Field, field_validator

BD_MOBILE_REGEX = re.compile(r"^01[3-9]\d{8}$")

def normalize_bd_phone(value: str | None):
    if value is None:
        return None

    phone = str(value).strip().replace(" ", "").replace("-", "")

    if not phone:
        return None

    if phone.startswith("+880"):
        phone = "0" + phone[4:]
    elif phone.startswith("880"):
        phone = "0" + phone[3:]

    if not BD_MOBILE_REGEX.fullmatch(phone):
        raise ValueError(
            "Phone number must be a valid Bangladesh mobile number. "
            "Example: 01712345678 or +8801712345678."
        )

    return phone

class EmployeeBase(BaseModel):
    employee_name: str = Field(..., max_length=255)

    official_employee_id: str | None = None

    photo_url: str | None = None
    photo_thumb_url: str | None = None
    passport_photo_url: str | None = None
    photo_original_name: str | None = None
    photo_mime_type: str | None = None
    photo_size_bytes: int | None = None
    signature_url: str | None = None

    email: EmailStr | None = None
    phone: str | None = None
    nid: str | None = None

    dob: date | None = None
    joining_date: date | None = None

    gender: str | None = None
    employee_type: str = "Permanent"

    company_id: int
    branch_id: int
    department_id: int
    designation_id: int

    reporting_to_employee_id: int | None = None

    remarks: str | None = None

    @field_validator("phone", mode="before")
    @classmethod
    def validate_phone(cls, value):
        return normalize_bd_phone(value)


class EmployeeCreate(EmployeeBase):
    employee_code: str | None = None


class EmployeeUpdate(BaseModel):
    employee_name: str | None = None
    employee_code: str | None = None

    official_employee_id: str | None = None

    photo_url: str | None = None
    photo_thumb_url: str | None = None
    passport_photo_url: str | None = None
    photo_original_name: str | None = None
    photo_mime_type: str | None = None
    photo_size_bytes: int | None = None
    signature_url: str | None = None

    email: EmailStr | None = None
    phone: str | None = None
    nid: str | None = None

    dob: date | None = None
    joining_date: date | None = None

    gender: str | None = None
    employee_type: str | None = None

    company_id: int | None = None
    branch_id: int | None = None
    department_id: int | None = None
    designation_id: int | None = None

    reporting_to_employee_id: int | None = None

    remarks: str | None = None
    is_active: bool | None = None

    @field_validator("phone", mode="before")
    @classmethod
    def validate_phone(cls, value):
        return normalize_bd_phone(value)


class EmployeeResponse(EmployeeBase):
    id: int
    employee_code: str

    is_active: bool

    company_name: str | None = None
    branch_name: str | None = None
    department_name: str | None = None
    designation_name: str | None = None

    reporting_to_employee_name: str | None = None

    created_by: str | None = None
    updated_by: str | None = None

    created_by_name: str | None = None
    updated_by_name: str | None = None

    created_at: datetime
    updated_at: datetime


class EmployeeListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[EmployeeResponse]


class EmployeeMessageResponse(BaseModel):
    message: str
    data: EmployeeResponse | None = None

File: frontend/src/types/employee.ts
// E:\Audit\AMS\frontend\src\types\employee.ts

export type Employee = {
  id: number;

  employee_code: string;
  official_employee_id: string | null;
  employee_name: string;

  photo_url: string | null;
  photo_thumb_url: string | null;
  signature_url: string | null;
  signature_thumb_url: string | null;

  email: string | null;
  phone: string | null;
  nid: string | null;

  dob: string | null;
  joining_date: string | null;
  gender: string | null;

  /**
   * Existing frontend form uses employee_type.
   * Backend response may use employee_status.
   * Keep both until naming is standardized.
   */
  employee_type: string | null;
  employee_status: string | null;

  company_id: number;
  branch_id: number;
  department_id: number;
  designation_id: number;
  reporting_to_employee_id: number | null;

  remarks: string | null;
  is_active: boolean;

  company_name: string | null;
  branch_name: string | null;
  department_name: string | null;
  designation_name: string | null;
  reporting_to_employee_name: string | null;

  created_by: string | null;
  updated_by: string | null;
  created_by_name: string | null;
  updated_by_name: string | null;

  created_at: string;
  updated_at: string;
};

export type EmployeeCreatePayload = {
  employee_code?: string | null;
  official_employee_id?: string | null;
  employee_name: string;

  email?: string | null;
  phone?: string | null;
  nid?: string | null;

  dob?: string | null;
  joining_date?: string | null;
  gender?: string | null;

  employee_type?: string | null;
  employee_status?: string | null;

  company_id: number;
  branch_id: number;
  department_id: number;
  designation_id: number;
  reporting_to_employee_id?: number | null;

  remarks?: string | null;
};

export type EmployeeUpdatePayload = Partial<EmployeeCreatePayload>;

export type EmployeeListResponse = {
  items: Employee[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
};

export type EmployeeMessageResponse = {
  message: string;
  data: Employee | null;
};